test/feature/dynamic_types/DynamicTypesTests.cpp

```diff
@@ -0,0 +1,7669 @@
+    EXPECT_EQ(md->index(), 0u);
+    EXPECT_EQ(md->name(), md1->name());
+    EXPECT_EQ(md->type(), md1->type());
+
+    dm1 = members_by_name["int64"];
+    ASSERT_EQ(RETCODE_OK, dm1->get_descriptor(md));
+    EXPECT_EQ(md->index(), 1u);
+    EXPECT_EQ(md->name(), md2->name());
+    EXPECT_EQ(md->type(), md2->type());
+
+    // • checking indexes work according with OMG standard 1.3 section 7.5.2.7.6
+    md = traits<MemberDescriptor>::make_shared();
+    md->id(7);
+    md->name("bool");
+    md->type(factory->get_primitive_type(eprosima::fastdds::dds::TK_BOOLEAN));
+    ASSERT_EQ(RETCODE_OK, struct_type_builder->add_member(md));
+
+    struct_type_builder->get_all_members(members_by_id);
+    ASSERT_EQ(3, members_by_id.size());
+
+    MemberDescriptor::_ref_type tmp = traits<MemberDescriptor>::make_shared();
+    auto dm = members_by_id[3];
+    ASSERT_EQ(RETCODE_OK, dm->get_descriptor(tmp));
+    EXPECT_EQ(tmp->index(), 0u);
+    EXPECT_EQ(tmp->name(), md1->name());
+    EXPECT_EQ(tmp->type(), md1->type());
+
+    dm = members_by_id[7];
+    ASSERT_EQ(RETCODE_OK, dm->get_descriptor(tmp));
+    EXPECT_EQ(tmp->name(), md->name());
+    EXPECT_EQ(tmp->type(), md->type());
+
+    dm = members_by_id[1];
+    ASSERT_EQ(RETCODE_OK, dm->get_descriptor(tmp));
+    EXPECT_EQ(tmp->index(), 1u);
+    EXPECT_EQ(tmp->name(), md2->name());
+    EXPECT_EQ(tmp->type(), md2->type());
+
+
+    // • checking adding duplicates
+    {
+        eprosima::fastdds::dds::Log::ScopeLogs _("disable");
+        //    + duplicate name
+        md = traits<MemberDescriptor>::make_shared();
+        md->id(1);
+        md->name("int32");
+        md->type(factory->get_primitive_type(eprosima::fastdds::dds::TK_INT32));
+        EXPECT_NE(RETCODE_OK, struct_type_builder->add_member(md));
+
+        //    + duplicate id
+        md = traits<MemberDescriptor>::make_shared();
+        md->id(7);
+        md->name("dup_bool");
+        md->type(factory->get_primitive_type(eprosima::fastdds::dds::TK_BOOLEAN));
+        EXPECT_NE(RETCODE_OK, struct_type_builder->add_member(md));
+    }
+}
+
+TEST_F(DynamicTypesTests, DynamicTypeBuilderFactory)
+{
+    DynamicTypeBuilderFactory::_ref_type factory {DynamicTypeBuilderFactory::get_instance()};
+
+    // Try to create with invalid values
+    // • strings
+    DynamicTypeBuilder::_ref_type created_builder {factory->create_string_type(static_cast<uint32_t>(LENGTH_UNLIMITED))};
+    ASSERT_TRUE(created_builder);
+
+    DynamicType::_ref_type type {created_builder->build()};
+    ASSERT_TRUE(type);
+    DynamicType::_ref_type type2 {created_builder->build()};
+    ASSERT_TRUE(type2);
+
+    ASSERT_TRUE(type->equals(type2));
+
+    // • wstrings
+    created_builder = factory->create_wstring_type(static_cast<uint32_t>(LENGTH_UNLIMITED));
+    ASSERT_TRUE(created_builder);
+
+    type = created_builder->build();
+    ASSERT_TRUE(type);
+    type2 = created_builder->build();
+    ASSERT_TRUE(type2);
+
+    ASSERT_TRUE(type->equals(type2));
```

Comment:
Looking at the `create_string_type`/`create_wstring_type` even though it is documented that it could return a nil reference, current implementation would never return that the descriptor is inconsistent. Consequently, in order to improve coverage, I suggest transforming the `is_consistent` check into an assertion and removing the possible nil reference return value from the documentation.

Reply:
Done in 2a37cf615